clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1850,23 +1857,41 @@ class SyclKernelDecompMarker : public SyclKernelFieldHandler {
       CollectionStack.back() = true;
       PointerStack.pop_back();
     } else if (PointerStack.pop_back_val()) {
-      // FIXME: Array of pointers/ array of type containing pointers
-      // will be handled in a follow up PR. Currently, they continue
-      // to trigger decomposition.
-      if (!FD->hasAttr<SYCLRequiresDecompositionAttr>())
-        FD->addAttr(SYCLRequiresDecompositionAttr::CreateImplicit(
-            SemaRef.getASTContext()));
-      CollectionStack.back() = true;
+      if (!FD->hasAttr<SYCLGenerateNewTypeAttr>())
```

Comment:
We check when adding SYCLRequiresDecompositionAttr that it does not have SYCLGenerateNewTypeAttr; perhaps we need a check here too.   At least an assertion.

Reply:
Can you elaborate please? We don't check for `SYCLGenerateNewTypeAttr` when adding `SYCLRequiresDecompositionAttr`. We add these attributes based on value in `CollectionStack `and `PointerStack`. We did the same thing when handling records in `leaveStruct`.  `CollectionStack` determines if we `addSYCLRequiresDecompositionAttr `and if added we don't enter the 'else' block when we add `SYCLGenerateNewTypeAttr`

Reply:
Oops!  Nvm, I misread the change-markers.